taichi/ui/common/canvas_base.h

```diff
@@ -0,0 +1,45 @@
+#pragma once
+#include "taichi/ui/common/field_info.h"
+#include "taichi/ui/common/scene_base.h"
+#include "taichi/ui/common/renderable_info.h"
+#include "taichi/ui/utils/utils.h"
+
+TI_UI_NAMESPACE_BEGIN
+
+struct SetImageInfo {
+  FieldInfo img;
+};
+
+struct TrianglesInfo {
+  RenderableInfo renderable_info;
+  glm::vec3 color;
+};
+
+struct CirclesInfo {
+  RenderableInfo renderable_info;
+  glm::vec3 color;
+  float radius;
+};
+
+struct LinesInfo {
+  RenderableInfo renderable_info;
+  glm::vec3 color;
+  float width;
+};
+
+struct CanvasBase {
+  virtual void set_background_color(const glm::vec3 &color) {
```

Comment:
1. Use `virtual ...() = 0;`?
2. You need to add a virtual destructor, e.g. `virtual ~CanvasBase() = default;`